kratos/processes/apply_ray_casting_process.cpp

```diff
@@ -0,0 +1,485 @@
+
+        // First clearing the intersections points vector
+        rIntersections.clear();
+
+		// Get the octree from the parent discontinuous distance process
+        OctreeType* pOctree = mpFindIntersectedObjectsProcess->GetOctreePointer();
+
+		// Compute the normalized ray key
+        OctreeType::key_type ray_key[3] = {pOctree->CalcKeyNormalized(ray[0]), pOctree->CalcKeyNormalized(ray[1]), pOctree->CalcKeyNormalized(ray[2])};
+
+        // Getting the entrance cell from lower extreme
+        OctreeType::key_type cell_key[3];
+        OctreeType::cell_type* cell = pOctree->pGetCell(ray_key);
+        while (cell) {
+			// Get the current cell intersections
+            const int cell_int = this->GetCellIntersections(cell, ray, ray_key, direction, rIntersections);
+			KRATOS_ERROR_IF(cell_int != 0)
+				<< "Error in GetCellIntersections for ray [" << ray[0] << "," << ray[1] << "," << ray[2] << "] with direction " << direction << std::endl;
+			// And if it exists, go to the next cell
+            if (cell->GetNeighbourKey(1 + direction * 2, cell_key)) {
+                ray_key[direction] = cell_key[direction];
+                cell = pOctree->pGetCell(ray_key);
+                ray_key[direction] -= 1 ; // The key returned by GetNeighbourKey is inside the cell (minkey +1), to ensure that the corresponding cell get in pGetCell is the right one.
+            } else {
+                cell = NULL;
+			}
+        }
+
+        // Eliminate the repeated intersecting objects
+        if (!rIntersections.empty()) {
+            // Sort
+            std::sort(rIntersections.begin(), rIntersections.end());
+            // Unique
+            std::vector<std::pair<double, Element::GeometryType*> >::iterator i_begin = rIntersections.begin();
+            std::vector<std::pair<double, Element::GeometryType*> >::iterator i_intersection = rIntersections.begin();
+            while (++i_begin != rIntersections.end()) {
+                // considering the very near points as the same points
+                if (std::abs(i_begin->first - i_intersection->first) > epsilon) // if the hit points are far enough they are not the same
+                    *(++i_intersection) = *i_begin;
+            }
+            rIntersections.resize((++i_intersection) - rIntersections.begin());
+        }
+	}
+
+	template<std::size_t TDim>
+	int  ApplyRayCastingProcess<TDim>::GetCellIntersections(
+		OctreeType::cell_type* cell,
+		const double* ray,
+		OctreeType::key_type* ray_key,
+		const unsigned int direction,
+		std::vector<std::pair<double, Element::GeometryType*> > &rIntersections)
+	{
+		//This function passes the ray through the cell and gives the hit point to all objects in its way
+		//ray is of dimension (3) normalized in (0,1)^3 space
+		// direction can be 0,1,2 which are x,y and z respectively
+		typedef OctreeType::cell_type::object_container_type object_container_type;
+		object_container_type* objects = (cell->pGetObjects());
```

Comment:
```suggestion
		auto objects = (cell->pGetObjects());
```
(if this works I think we can remove the `typedef` above)

Reply:
Actually the `typedef` is needed in a loop below if we don't change it to an auto loop.